packages/snaps-controllers/src/snaps/SnapController.ts

```diff
@@ -1438,6 +1451,77 @@ export class SnapController extends BaseController<
     return encodeAuxiliaryFile(value, encoding);
   }
 
+  /**
+   * Processes dynamic permission request.
+   *
+   * @param snapId - The id of the Snap requesting permissions.
+   * @param permissions - Structure with requested permissions.
+   * @returns Result of permission grant request.
+   */
+  async requestPermissions(
+    snapId: SnapId,
+    permissions: SnapPermissions,
+  ): Promise<SubjectPermissions<PermissionConstraint>> {
+    log(`Granting dynamic permissions to the Snap: ${snapId}`);
+    const snapsState = this.state.snaps;
+    const snap = snapsState[snapId];
+    const { dynamicPermissions } = snap; // Use to validate
+
+    if (!dynamicPermissions) {
+      throw new Error('This Snap have no dynamic permissions specified.');
+    }
+
+    const requestedPermissionsKeys = Object.keys(permissions);
+    const dynamicPermissionKeys = Object.keys(dynamicPermissions);
+
+    // Ensure that requested dynamic permissions are allowed
```

Comment:
Aside from matching the input with the dynamic permission keys specified in the manifest we also need to check that caveats match exactly.

Reply:
Added caveats check as well and refactored the rest.